ipa-core/src/query/runner/hybrid.rs

```diff
@@ -1,37 +1,285 @@
-use std::{marker::PhantomData, sync::Arc};
+use std::{
+    collections::HashSet,
+    marker::PhantomData,
+    sync::{Arc, Mutex},
+};
+
+use futures::{stream::iter, StreamExt, TryStreamExt};
 
 use crate::{
     error::Error,
+    ff::boolean_array::{BA20, BA3, BA8},
     helpers::{
         query::{HybridQueryParams, QuerySize},
-        BodyStream,
+        BodyStream, LengthDelimitedStream,
     },
     hpke::PrivateKeyRegistry,
+    protocol::{context::UpgradableContext, ipa_prf::shuffle::Shuffle, step::ProtocolStep::Hybrid},
+    report::{hybrid::HybridReport, EncryptedOprfReport},
     secret_sharing::{replicated::semi_honest::AdditiveShare as ReplicatedShare, SharedValue},
 };
 
 pub struct Query<C, HV, R: PrivateKeyRegistry> {
-    _config: HybridQueryParams,
-    _key_registry: Arc<R>,
+    config: HybridQueryParams,
+    key_registry: Arc<R>,
     phantom_data: PhantomData<(C, HV)>,
 }
 
-impl<C, HV: SharedValue, R: PrivateKeyRegistry> Query<C, HV, R> {
+impl<C, HV: SharedValue, R: PrivateKeyRegistry> Query<C, HV, R>
+where
+    C: UpgradableContext + Shuffle,
+{
     pub fn new(query_params: HybridQueryParams, key_registry: Arc<R>) -> Self {
         Self {
-            _config: query_params,
-            _key_registry: key_registry,
+            config: query_params,
+            key_registry,
             phantom_data: PhantomData,
         }
     }
 
     #[tracing::instrument("hybrid_query", skip_all, fields(sz=%query_size))]
     pub async fn execute(
         self,
-        _ctx: C,
+        ctx: C,
         query_size: QuerySize,
-        _input_stream: BodyStream,
+        input_stream: BodyStream,
     ) -> Result<Vec<ReplicatedShare<HV>>, Error> {
-        unimplemented!()
+        let Self {
+            config,
+            key_registry,
+            phantom_data: _,
+        } = self;
+        tracing::info!("New hybrid query: {config:?}");
+        let _ctx = ctx.narrow(&Hybrid);
+        let sz = usize::from(query_size);
+
+        let seen_encrypted_reports: Arc<Mutex<HashSet<Vec<u8>>>> =
```

Comment:
I suggest to build a separate struct to check for duplicates. It may provide better encapsulation and it could work with reports, rather than just bytes. There is also a benefit of testing it more easily, in isolation with `execute` method

```rust
struct Duplicates(HashSet<...>);

impl Duplicates {

   fn check_duplicate(&mut self, report: EncryptedReport<>) -> Result<(), Duplicate>> {}

   // this should tell you the index of duplicate, if it is there
   fn check_duplicates(&mut self, reports: &[EncryptedReport]) -> Result<(), DuplicateReportIndex> {}
}
```

Allowing it to operate on a vec, would help `execute` function too, as it no longer needs to dance around `map_ok`.

Reply:
this makes sense, though it doesn't seem that HashSets are ordered, so there isn't an index. I think it's sufficient to just return an Error.

Reply:
yea, I meant the index in the `reports` slice  - you want to know which exact report is a duplicate